Major Overhaul: Scope of Core & Simplification (#232)
ietf-rats-wg/draft-ietf-rats-corim: draft-ietf-rats-corim.md

| Series Endorsements | List of expected actual state claims, Series of tuples containing selection and series Endorsement claims | If the list of expected claims are in the ACS, and if the series selection condition is satisfied, then add the series Endorsement claims for the matching series tuple to the ACS with Endorser authority |
| Verifier | List of expected actual state claims, List of Verifier generated claims | If the list of expected claims are in the ACS, then add the list of Verifier generated claims to the ACS with Verifier authority |
| Policy | List of expected actual state claims, List of Policy generated claims | If the list of expected claims are in the ACS, then add the list of Policy generated claims to the ACS with Policy Owner authority |
| Attestation Results | List of expected actual state claims, List of expected Attestation Results claims | If the list of expected claims are in the ACS, then copy the list of Attestation Results claims into the ARS |
{: #tbl-cmrr title="Conceptual Message Representation Requirements"}

## Quantizing Inputs {#sec-quantize}

> Inputs may change soon after a Verifier processes them causing the Verifier to consider how often it should check for updated inputs.

Reviewer: I agree with @deeglaze

This document specifies Concise Reference Integrity Manifests (CoRIM)–a CBOR {{-cbor}} based data model addressing the above challenges by using an extensible format common to all supply chain actors and Verifiers.
CoRIM enables Verifiers to reconcile a complex distributed supply chain into a single homogeneous view. See {{sec-verifier-rec}}.

## Terminology and Requirements Language

This document uses terms and concepts defined by the RATS architecture.
For a complete glossary see {{Section 4 of -rats-arch}}.

In this document, the term CoRIM message and CoRIM documents are used as synonyms. A CoRIM data structure can be at rest (e.g., residing in a file system as a document) or can be in flight (e.g., conveyed as a message in a protocol exchange). The bytes composing the CoRIM data structure are the same either way.

The terminology from CBOR {{-cbor}}, CDDL {{-cddl}} and COSE {{-cose}} applies;
in particular, CBOR diagnostic notation is defined in {{Section 8 of -cbor}}
and {{Section G of -cddl}}. Terms and concepts are always referenced as proper nouns, i.e., with Capital Letters.

{::boilerplate bcp14}

# Verifier Reconciliation {#sec-verifier-rec}

> While this specification focuses on the CoRIM format itself, it is vital to understand its intended meaning for the mechanics of how Verifiers should process CoRIM documents.

Reviewer: Thanks for the attempt to clarify the sentence.

However, the document focuses on the description of the data format & the processing.  There is no primary vs secondary focus.

> While this specification focuses on the CoRIM format itself, it is vital to understand its intended meaning for the mechanics of how Verifiers should process CoRIM documents.

Reviewer: The corrected text fixes my previous concern.

A Verifier has to reconcile its various inputs, CoRIM being one of them.
Complementary to the external CoRIM documents, a Verifier is assumed to construct an internal representation of each input and map each external representation to an internal representation.
Using the internal representation, the Verifier processes inputs as though they are parts of a conversation where the Verifier keeps track of who said what.
Namely, the provenance of the inputs is tracked as *authority*.

> In the case of CoRIM, the CoRIM issuer identifies its authority.

Reviewer 2: We agreed to keep the older version, of the text as there is no permanent external representation and New text 167 creates that `ambiguity`

> In the case of CoRIM, the CoRIM issuer identifies its authority.

Reviewer: which "older version" are you referring to?

Both OLD and NEW have the same "external"...

Effectively, Attesters, Reference Value Providers, Endorsers, Verifier Owners, Relying Parties, and even the Verifier potentially all contribute to the conversation.
Each producer of corresponding RATS Conceptual Messages can assert Claims about an Attester's actual or allowed state.
The Verifier's objective is to produce a list of Claims that describe the Attester's presumed actual state.
Producers of RATS Conceptual Messages can assert contradictory assertions.

> For example, a compromised Attester may produce false assertions that create a conflict with an RVP's issued Reference Values.

Reviewer: ```suggestion
For example, a compromised Attester may produce false statements that conflict with the Reference Values provided by a Reference Value Provider (RVP).
```

> In essence, if Evidence is not corroborated by a Reference Value Provider's (RVP) assertions, then the RVP's Claims are not included in the internal representation of an Attester's actual state, a.k.a., Accepted Claims Set (ACS).

Reviewer: ```suggestion
In essence, if Evidence is not corroborated by an RVP's assertions, then the RVP's Claims are not included in the internal representation of an Attester's actual state, a.k.a., Accepted Claims Set (ACS).
```

> *[Ned] Suggest we use Actual-state Claims Set (ACS) since "accepted" implies an acceptance policy input.*

A Verifier relies on input from appraisal policy to identify relevant assertions included in the ACS.
For example, if a policy requires corroborated assertions issued by a particular RVP, then those assertions may be conveyed as Attestation Results.
The Verifier may produce new assertions as a result of an applied appraisal policy.
For example, if an appraisal procedure finds all of the components of a subsystem are configured correctly, the policy may direct the Verifier to produce new assertions, "Subsystem=X" has the Claim "TRUSTED=TRUE".
Consequently, the internal ACS structure is a reconciled conversation between several producers of RATS Conceptual Messages that has mapped each message into a consistent internal representation, has associated the identity of the corresponding RATS role with each assertion (the authority), and has applied Conceptual Message constraints to the assertion.

> The CoRIM data model specified in this document covers the RATS Conceptual Message types, `Reference Values` and `Endorsements`.

Reviewer: ```suggestion
The CoRIM data model specified in this document covers the RATS Conceptual Message types, "Reference Values" and "Endorsements".
```

> Reference values and Endorsements are required for Verifier reconciliation (and Evidence is required for corresponding internal ACS creation) as illustrated in {{sec-interact-acs}}.

Reviewer: ```suggestion
Reference values and Endorsements are required for Verifier reconciliation, and Evidence is required for corresponding internal ACS creation as illustrated in {{sec-interact-acs}}.
```

## Internal Representation {#sec-internal-rep}

In this document CDDL is used to specify both the CoRIM structure and to specify an internal representation for use in the appraisal procedure.
The actual internal representation of a Verifier is implementation-specific and out-of-scope of this document.
Requirements for an internal representation of Conceptual Messages are defined in {{tbl-cmrr}}, where each Conceptual Message type has a structure as depicted by the *Structure* column.
The internal representations used by this document are defined in {{sec-ir-cm}}.

## Interacting with an ACS {#sec-interact-acs}

Conceptual Messages interact with an ACS by specifying criteria that should be met by the ACS and by presenting the assertions that should be added to the ACS if the criteria are satisfied.

> Internal representations of Conceptual Messages, ACS, and ARS should satisfy the following requirements for Verifier reconciliation and appraisal processing:

Reviewer: need to expand "ARS" on first use :-)

> Internal representations of Conceptual Messages, ACS, and ARS should satisfy the following requirements for Verifier reconciliation and appraisal processing:

Reviewer 2: ```suggestion
Internal representations of Conceptual Messages, ACS, and Attestation Results Set (ARS) should satisfy the following requirements for Verifier reconciliation and appraisal processing:
```

Reviewer 2: Add a reference to ARS

| CM Type | Structure | Description |
|---
| Evidence | List of Evidence claims | If the Attester is authenticated, add Evidence claims to the ACS with Attester authority |
| Reference Values | List of Reference Values claims | If a reference value in a CoRIM matches claims in the ACS, then the authority of the CoRIM issuer is added to those claims. |
| Endorsements | List of expected actual state claims, List of Endorsed Values claims | If the list of expected claims are in the ACS, then add the list of Endorsed Values claims to the ACS with Endorser authority |

> | Series Endorsements | List of expected actual state claims, Series of tuples containing selection and series Endorsement claims | If the list of expected claims are in the ACS, and if the series selection condition is satisfied, then add the series Endorsement claims for the first series tuple with satisfied condition to the ACS with Endorser authority |

Reviewer: unclear what "selection" means in this context

Reviewer 3: It is a second stage condition. 
The series structure is a list of condition-addition 2-tuples. Selection is another word for condition that isn't confused with the endorsement's first-stage condition.
Possibly, the second sentence here is a clearer explanation of selection. The current text assumes the reader read ahead to understand it's use here.

Reviewer: OK, thanks. The problem is this is not self-contained.  At least there should be a forward reference.

> | Series Endorsements | List of expected actual state claims, Series of tuples containing selection and series Endorsement claims | If the list of expected claims are in the ACS, and if the series selection condition is satisfied, then add the series Endorsement claims for the first series tuple with satisfied condition to the ACS with Endorser authority |

Reviewer 3: ```suggestion
| Series Endorsements | List of expected actual state claims and a series of selection-addition tuples | If the expected claims are in the ACS, and if the series selection condition is satisfied, then add the additional claims to the ACS with Endorser authority. See {{sec-ir-end-val}} |
```

Reviewer: nice, thanks!

| Verifier | List of expected actual state claims, List of Verifier-generated claims | If the list of expected claims are in the ACS, then add the list of Verifier-generated claims to the ACS with Verifier authority |
| Policy | List of expected actual state claims, List of Policy-generated claims | If the list of expected claims are in the ACS, then add the list of Policy-generated claims to the ACS with Policy Owner authority |
| Attestation Results | List of expected actual state claims, List of expected Attestation Results claims | If the list of expected claims are in the ACS, then copy the list of Attestation Results claims into the ARS |
{: #tbl-cmrr title="Conceptual Message Representation Requirements"}

## Quantizing Inputs {#sec-quantize}

> To be completed!

Reviewer: add a reference to the tracking issue instead

Reviewer 2: issue #242

> Inputs may change soon after a Verifier processes them causing the Verifier to consider how often it should check for updated inputs.

Reviewer 2: ```suggestion

```

Besides, one supply chain actor will only provide the subset of characteristics that they know about the Attester; typically, because a certain supply chain actor is the responsible authority or because the system boundaries of the described characteristics is defined by a certain system component/module.
Attesters vary from one vendor to another and for a given vendor from one product to another.
Not only Attesters can evolve and therefore new measurement types need to be expressed, but an Endorser may also want to provide new security relevant attributes about an Attester at a future point in time.

This document specifies Concise Reference Integrity Manifests (CoRIM); a CBOR {{-cbor}} based data model addressing the above challenges by using an extensible format common to all supply chain actors and Verifiers.
CoRIM enables Verifiers to reconcile a complex and scattered supply chain into a single homogeneous view.

## Verifier Reconciliation {#sec-verifier-rec}

While this specification focusses on the CoRIM format itself, understanding the procedures how Verifiers are assumed to digest CoRIM documents provides vital context.

> A Verifier hast to reconcile its various inputs, CoRIM being one of them.

Author: ```suggestion
A Verifier has to reconcile its various inputs, CoRIM being one of them.
```

~~~ cddl
ECT = non-empty<{
  ? &(context: 0) => text
  ? &(environment: 1) => environment-map
  ? &(claims: 2) => claims-map / [ + local-ECT ]
  ? &(authority: 3) => [ + $cryptokeytypechoice ]
}>
~~~

Context (label 0):

: A text string that TBD (todo: specify here)

Environment (label 1):

: A set of claims that identifies instances or classes of environments which are part of a composite Attester's components.

Claims (label 2):

: A set of claims with semantics that depend on the context in which an ECT is used.
In external CoRIM representations, claims can either represent Reference Values, Endorsed Values, or Conditions that have to be met.
To illustrate examples of internal representations, ECTs are used to represent ACS.

Authority (label 3):

: A certain type of key material by which the authority (and corresponding provenance) of an ECT can be determined, such as the public key of an asymmetric key pair that is associated with an authority's PKIX certificate.

## Internal Representation {#sec-internal-rep}

An ACS is a list of ECTs.

~~~ cddl
ACS = {
  &(actual-state: 0) => [ + ECT ]
}
~~~

> In this document, CDDL definitions are used to illustrate examples for internal representations.

Author: Isn't this going a bit far? This document definitely comes with CDDL for CoRIM in CBOR, right?

The type of the digest algorithm identifier can be either `int` or `text` and is interpreted according to the {{-named-info}} registry.
Specifically, `int` values are matched against "ID" entries, `text` values are matched against "Hash Name String" entries.
Whenever possible, using the `int` encoding is RECOMMENDED.

~~~ cddl
{::include cddl/digest.cddl}
~~~

A measurement can be obtained using different hash algorithms.
A `digests-type` can be used to collect multiple digest values obtained by applying different hash algorithms on the same input.
Each entry in the `digests-type` MUST have a unique `alg` value.

## Tagged Bytes Type {#sec-common-tagged-bytes}

An opaque, variable-length byte string.
It can be used in different contexts: as an instance, class or group identifier in an `environment-map`; as a raw value measurement in a `measurement-values-map`.
Its semantics are defined by the context in which it is found, and by the overarching CoRIM profile.
When used as an identifier the responsible allocator entity SHOULD ensure uniqueness within the context that it is used.

~~~ cddl
{::include cddl/tagged-bytes.cddl}
~~~

> # Appraisal of CoRIM-based Inputs {#sec-appr-corim-inputs}

Author: CMW is an umbrella structure ofr all types of RATS Conceptual Messages. You could - I am making things up here - bundle TPM Evidence, EAT Attestation Results, and a CoRIM in a CMW (each with their own media type).

Inputs to a Verifier are mapped from their external representation to an internal representation.
CoRIM defines structure for Conceptual Messages that includes Endorsements and Reference Values.

> CoRIM data structures may be used by Evidence and Attestation Results that wish to describe overlapping structure.

Author: I do not want to loose this concern/comment when we merge the PR. @deeglaze Do you want to rescue this comment into a better visible issue?

CoRIM-based data structures define an external representation of Conceptual Messages that are mapped to an internal representation.
Appraisal processing describes both mapping transformations and Verifier reconciliation {{sec-verifier-rec}}.
Non-CoRIM-based data structures require mapping transformation, but these are out of scope for this document.

If a CoRIM profile is specified, there are a few well-defined points in the procedure where Verifier behaviour

> depends on the profile.

Author: fixed

## Appraisal Procedure {#sec-appraisal-procedure}

The appraisal procedure is divided into several logical phases.

### **Phase 1**: Input Validation and Transformation

During Phase 1, Conceptual Message inputs are cryptographically validated, such as checking digital signatures.
Inputs are transformed from their exernal representations to an internal representation.
Internal representations are staged for appraisal processing, such as populating an input queue.

### **Phase 2**: Evidence Augmentation

During Phase 2, Evidence inputs are added to a list that describes the Attester's actual state.
These inputs are added with the Attester's authority.

### **Phase 3**: Reference Values Corroboration and Augmentation

During Phase 3, Reference Values inputs are compared with Evidence inputs.
Reference Values inputs describe possible states of Attesters.
If the actual state of the Attester is described by the possible Attester states, then the overlapping (corroborated) actual states are added to the Attester's actual state.
These inputs are added with the Reference Value Provider's authority.

### **Phase 4**: Endorsed Values Augmentation

> During Phase 4, Endorsed Values inputs containing conditions that describe expected Attester state are compared.

Reviewer 2: Compared against whom ?

Author: Addressed in [4857331](https://github.com/ietf-rats-wg/draft-ietf-rats-corim/pull/232/commits/48573314c87a73fda94200a755c207e2efd21ec5)

### **Phase 4**: Endorsed Values Augmentation

During Phase 4, Endorsed Values inputs containing conditions that describe expected Attester state are compared.

> If the comparison is satisfied, then additional Claims about the Attester are added to the Accepted Claim Set.

Author: ```suggestion
If the comparison is satisfied, then additional Claims about the Attester are added to the Actual-State Claim Set.
```

Reviewer 2: Bulk of the document states ACS as Accepted Claims Set.  We need to agree on a common term and replace the entire document for consistency!

Reviewer 4: Indeed "actual" is more philosophical. We're only accepting an authority's attestation of actual state. Ceci n'est pas une pipe.

Reviewer 3: I'm OK with "Actual-state Claims Set" but it should be applied consistently. I prefer using ACS everywhere and define ACS in one place. Then it's easy to make a global change.

>
> * How cryptographic verification key material is represented (e.g., using Attestation Keys triples, or CoTS tags)
> * How key material is associated with the Attesting Environment
> * How the Attesting Environment is identified in Evidence

### Evidence Tranformation to Internal Represenatation

Evidence is divided up into one or more `ev` relations where the `condition` ECT identifies the Attester from which Evidence was collected. If the Verifier maintains multiple Attester sessions, the Verifier session may be identified using an ECT.

Evidence information is mapped to an `addition` ECT that populates each of the ECT fields. If the Evidence doesn't have a value for the mandatory fields, the Verifier MUST NOT process the Evidence.

The `e` field is populated with a Target Environment identifier.
The `c` field is populated with the measurements collected by an Attesting Environment.
The `a` field is populated with the identity of the entity that asserted (e.g., signed) the Evidence.
The `ns` field is populated with the namespace context if supplied. For example, the Attester's manufacturer may have a URI that identifies the manufacturing series, family or architecture.

> Evidence transformation algorithms may be well-known, see {{sec-spdm}} and {{sec-dice}};

Author: no, not a cause for a line break
in simple kramdown `;` is just a character used in sentences

may be defined by a CoRIM profile ({{sec-corim-profile-types}});

> or may be supplied dynamically. The handling of dynamic Evidence transformation algorithms is out of scope for this document.

Author: fixed

## Evidence Appraisal {#sec-ev-appraisal}

### The Acepted Claims Set

At the end of the Evidence collection process Evidence has been converted into

> an internal represenetation suitable for appraisal. See {{sec-ir-cm}}.

Author: fixed

Verifiers are not required to use this as their internal representation,

> for the purposes of this document appraisal is described in terms of the above cited internal representation.

Author: fixed

Instead, if an entry is found, the digest comparison proceeds as defined in {{sec-cmp-digests}} after equivalence has been found according to {{sec-comid-integrity-registers}}.
Note that it is not required for all the entries in Evidence to be used during matching: the Reference Value could consist of a subset of the device's register space. In TPM parlance, a TPM "quote" may report all PCRs in Evidence, while a Reference Value could describe a subset of PCRs.

### Authority Comparison  {#sec-compare-auth}

The `a` field comparison tests for trust path termination.
If the authority of the first ECT is a trust anchor for the authority of the second ECT, the second ECT is valid.
If the authority values are identical, then the second ECT is valid.

### Name Space Comparison  {#sec-compare-ns}

The `ns` field comparison tests equality where the text values are identical.

### Conceptual Message Type Comparison  {#sec-compare-cm}

The `cm` field comparison tests equality of one or more bits.

### Profile-directed Comparison {#sec-compare-profile}

> A profile may specify handling for new CBOR tagged Reference Values. The

Reviewer 4: ```suggestion
A profile may specify handling for new CBOR tagged Reference Values.
The profile must specify how to compare the CBOR tagged Reference Value against the ACS.
```

--- back

# Transforming SPDM Evidence {#sec-spdm}

This section defines how Evidence from SPDM {{SPDM}} is transformed into a

> format where it can be added to an accepted claims set.

Author: fixed

A Verifier supporting SPDM format Evidence should implement this section.

The TCG DICE Concise Evidence Binding for SPDM specification {{CE.SPDM}} describes the process by which

> measurements in an SPDM Measurement Block are converted to Evidence suitable for

Author: fixed

matching using the rules below.
The SPDM measurements are converted to `concise-evidence` which has a format that is similar

> to CoRIM `triples-map` (their semantics follows the matching rules described above).

Author: fixed

# Transforming DICE Evidence {#sec-dice}

This section defines how Evidence from DICE {{DICE.AA}} is transformed into a

> format where it can be added to an accepted claims set.

Author: fixed

A Verifier supporting DICE format Evidence should implement this section.

DICE Evidence appears in certificates in the TcbInfo or MultiTcbInfo extension.
Each TcbInfo, and each entry in the MultiTcbInfo, is converted to an

> `endorsed-triple-record` using the rules in this section.

Author: fixed

In a MultiTcbInfo each entry in the sequence is treated as independent and

> translated into a separate Evidence object.

Author: fixed

> translated into a separate Evidence object.

Author: fixed

The Verifier SHALL translate each field in the TcbInfo into a field in the

> created endorsed-triple-record

Author: fixed

- The TcbInfo `vendor` field SHALL be copied to the field named `environment-map / class / vendor`
- The TcbInfo `model` field SHALL be copied to the field named `environment-map / class / model`
- The TcbInfo `layer` field SHALL be copied to the field named `environment-map / class / layer`
- The TcbInfo `index` field SHALL be copied to the field named `environment-map / class / index`

- The TcbInfo `version` field SHALL be translated to the field named `measurement-map / mval / version / version`
- The TcbInfo `svn` field SHALL be copied to the field named `measurement-map / mval / svn`
- The TcbInfo `fwids` field SHALL be translated to the field named `measurement-map / mval / digests`
  - Each digest within fwids is translated to a CoMID digest object, with an appropriate algorithm identifier
- The TcbInfo `flags` field SHALL be translated to the field named `measurement-map / mval / flags`
  - Each flag is translated independently
- The TcbInfo `vendorInfo` SHALL shall be copied to the field named `measurement-map / mval / raw-value`

If there are multiple `endorsed-triple-record`s with the same `environment-map`

> then they MUST be merged into a single entry.

Author: fixed

If the `measurement-values-map` fields in Evidence triples have conflicting

> values then the Verifier MUST fail validation.

Author: fixed
